tsdb/db.go

```diff
@@ -118,6 +118,10 @@ type Options struct {
 	// SeriesLifecycleCallback specifies a list of callbacks that will be called during a lifecycle of a series.
 	// It is always a no-op in Prometheus and mainly meant for external users who import TSDB.
 	SeriesLifecycleCallback SeriesLifecycleCallback
+
+	// MinValidTimeGracePeriod is the additional grace period w.r.t the current min valid time
```

Comment:
Why not have it as a hard value? Having the value consistently set to 1h for Prometheus would probably avoid some confusion.

Reply:
If with a hard value you mean a constant, there might not be a single value that satisfies all use cases for external users, so I think configurable is better. And I was thinking of this being 0 for Prometheus for now as we have a flag to change the block size (though hidden), and setting this to a fixed size means the data is no longer assured to be non-overlapping.

Reply:
I mean that rather as relative to whatever the current min block time, have it relative to the max block time.

Reply:
I'm also strongly opposed to this being configurable for Prometheus - we already have enough problems with users messing with the min block size and trying to use Prometheus for push without forcing Cortex design decisions onto Prometheus.

Reply:
Do you mean to say take in a multiplier and use min/max block time on that to get the grace period?

Having it >0 for Prometheus means that we need to enable vertical compaction always, so let it be 0 for Prometheus. This option is only exposed in the TSDB package but nowhere configurable from the Prometheus binary.

Reply:
I'm suggesting that rather than the head taking in 1-3h old data in Prometheus depending on where it is in the compaction cycle, that it'd always be a consistent 1h.

Reply:
I miss spoke in the issue description (fixed now). It is actually 1h old consistently for Prometheus. So with this grace period, it will be `1h + grace period` consistently.

The relevant piece of code which maintains that 1h: https://github.com/prometheus/prometheus/blob/7b24bb311631f4eaa5d794b849770f436fc75b5a/tsdb/head.go#L1015

Reply:
Ah, I didn't realise that.

In that case does that do the right thing in terms of blocks/WAL given that this minValidTime isn't tied to them?

Reply:
What do you mean by the right thing here? If you mean `1h + grace period` old for samples, then yes, it will be that. And during WAL replay, as we use the time passed in the `Head.Init` as the min valid time, we just allow upto that `mint - grace period` for samples.

Reply:
If the source data is coming from a Prometheus's WAL then that will be aligned with the 2h of the compaction cycle. This PR as proposed however is aligned with the current time. So there's a mismatch there.

Reply:
Nice catch. IIUC we want the out of bounds to be always 1h consistent. So I will fix it during replay to consider that.

Reply:
(after the open discussions are concluded in https://github.com/prometheus/prometheus/issues/7396)

Reply:
Hmm, that doesn't sound right if your goal is to handle delayed data from remote write as that's not 1h delayed - it goes back to the start of the head which is on the 2h cycle.

Reply:
I might be confusing things here. Here is what I understand right now:

1. Configurable grace period and not a fixed grace period. So, samples can overlap blocks too.
2. During an append, currently, it's fixed to max 1h old sample w.r.t. current maxt. So, this makes it `1h + grace period`
3. During WAL replay, we take min valid time as the maxt of the last block throughout the replay. So, while applying the grace period, we should take the min valid time w.r.t. the maxt of head block so that we don't end up with samples which are already in the block.

Reply:
That all sounds right, but that's not what I'm talking about.

Consider you have a normal Prometheus using remote write, which has remote write delayed for some reason and then restarts. It will start re-sending samples from the WAL based on where the head is.

Now that's going to Cortex with all this code, will that Cortex be able to ingest all the samples from that Prometheus's remote write and if so how much grace does it need to do so?

Reply:
> Now that's going to Cortex with all this code, will that Cortex be able to ingest all the samples from that Prometheus's remote write and if so how much grace does it need to do so?

Cortex will have a limit too on how old samples can it take and can't get away from it. The grace period will be a decision that needs to be taken on how long do you want to allow. This configuration would just give some extra time to fix things and wont completely eliminate losing data.